mlir/test/Transforms/test-operation-folder.mlir

```diff
@@ -34,3 +37,37 @@ func.func @test_dont_reorder_constants() -> (i32, i32, i32) {
   %2 = arith.constant 3 : i32
   return %0, %1, %2 : i32, i32, i32
 }
+
+func.func @test_dont_fold() -> (i32, i32, i32, i32, i32, i32) {
+  // Test either not folding or deduping constants.
+
+  // CHECK-LABEL: test_dont_fold
+  // CHECK-NOT: arith.constant 0
+  // CHECK-DAG: %[[CST:.+]] = arith.constant 0
+  // CHECK-DAG: %[[CST:.+]] = arith.constant 1
+  // CHECK-DAG: %[[CST:.+]] = arith.constant 2
+  // CHECK-DAG: %[[CST:.+]] = arith.constant 3
+  // CHECK-NEXT: return
+
+  // NOCSE-LABEL: test_dont_fold
+  // NOCSE-DAG: arith.constant 0 : i32
+  // NOCSE-DAG: arith.constant 1 : i32
+  // NOCSE-DAG: arith.constant 2 : i32
+  // NOCSE-DAG: arith.constant 1 : i32
+  // NOCSE-DAG: arith.constant 2 : i32
+  // NOCSE-DAG: arith.constant 3 : i32
```

Comment:
Is this really the behavior we want here? Seems like we still have some constant folding (more constants are introduced) but they aren't deduped!

Reply:
I was confused for second too :) This is the no CSE case. So we do have folding, just not constant depupe. This is flag Matthias suggested, so these two features are independent.

Reply:
The comment for fold says:

```
 /// Whether this should fold while greedily rewriting. This also disables
   /// CSE'ing constants.
   bool fold = true;
```

Why aren't we keeping these independent then? Disabling Fold shouldn't disable CSE if the CSE option is true...

Reply:
Out of date comment, sorry. It is/should be independent. I'll double check now.